qiita_db/user.py

```diff
@@ -126,48 +126,56 @@ def login(cls, email, password):
         IncorrectPasswordError
             Password passed is not correct for user
         """
-        # see if user exists
-        if not cls.exists(email):
-            raise IncorrectEmailError("Email not valid: %s" % email)
+        trans = Transaction("login_%s" % email)
 
-        if not validate_password(password):
-            raise IncorrectPasswordError("Password not valid!")
+        with trans:
+            # see if user exists
+            if not cls.exists(email, trans=trans):
+                raise IncorrectEmailError("Email not valid: %s" % email)
 
-        # pull password out of database
-        conn_handler = SQLConnectionHandler()
-        sql = ("SELECT password, user_level_id FROM qiita.{0} WHERE "
-               "email = %s".format(cls._table))
-        info = conn_handler.execute_fetchone(sql, (email, ))
-
-        # verify user email verification
-        # MAGIC NUMBER 5 = unverified email
-        if int(info[1]) == 5:
-            return False
-
-        # verify password
-        dbpass = info[0]
-        hashed = hash_password(password, dbpass)
-        if hashed == dbpass:
-            return cls(email)
-        else:
-            raise IncorrectPasswordError("Password not valid!")
+            if not validate_password(password):
+                raise IncorrectPasswordError("Password not valid!")
+
+            # pull password out of database
+            sql = ("SELECT password, user_level_id FROM qiita.{0} WHERE "
+                   "email = %s".format(cls._table))
+            trans.add(sql, [email])
+
+            info = trans.execute()[-1][0]
+
+            # verify user email verification
+            # MAGIC NUMBER 5 = unverified email
+            if int(info[1]) == 5:
```

Comment:
I know is not part of your code but can we use the function that retrieves the id from the value?